### PR TITLE
Adds the support for python 3

### PR DESCRIPTION
**Pull request by duxan**
Really nice library! Hope this one settles the support for python 3.

Tested with existing tests from test/test.py.

### CODE REVIEW

**Comment by RevolutionTech** [comment]
Tests work for me in Python 2, but when I try to run tests on Python 3.4 or Python 3.5, I get a segmentation fault:
```
$ python -V
Python 3.5.4
$ python test.py 
test_dl (__main__.TestClev) ... Segmentation fault
```

It will be difficult to diagnose what is happening on my local machine, so I will try to improve our CI setup a little bit so that we can run tests on a consistent environment.

**Comment by duxan** [comment]
I have tested under python 3.6, but tried fresh conda ENV with python 3.4.5 just now and it works under OS X Sierra (installed with `pip install git+https://github.com/duxan/weighted-levenshtein.git`)

image

**Comment by RevolutionTech** [comment]
Okay, cool. It's probably an issue with something in my local environment then. I've updated the repo to use `tox` so now we can add support for multiple Python versions by simply adding more python versions to the `envlist`.

Would you mind merging master into your branch and then adding the supported Python versions into the `tox.ini` so that tests run on all of these versions?

**Comment by duxan** [comment]
Sorry for the delay. I've added py3 to tox after some issues with conda and virtalenv working together (https://github.com/conda/conda/issues/1367). All tests are successful when run locally via tox for py27 and py36.

File: .gitignore [hunk 1]
```diff
@@ -35,6 +35,7 @@ __pycache__/
 
 # C extensions
 *.so
+*.c
```
Comment (RevolutionTech) [inline .gitignore]: I see why you added this, but maybe it would be better to add another `.gitignore` in the `weighted_levenshtein/` directory with just the line `clev.c`.

This `.gitignore` is automatically generated via gitignore.io so that one line change could be easily missed in the future if someone were to regenerate the file, plus ignoring all `.c` files might not be what we want in the future so I think this would be clearer.

File: setup.py [hunk 2]
```diff
@@ -2,18 +2,9 @@
...
-except ImportError:
-    extensions = [Extension("weighted_levenshtein.clev", ['weighted_levenshtein/clev.c'])]
-else:
-    extensions = cythonize([Extension("weighted_levenshtein.clev", ['weighted_levenshtein/clev.pyx'])])
-
-
 with open('README.rst') as readme:
     long_description = readme.read()
 
-
```
Comment (RevolutionTech) [inline setup.py]: nit: Please don't remove this empty newline

File: setup.py [hunk 3]
```diff
@@ -34,6 +25,8 @@
         'Programming Language :: Cython',
         'Programming Language :: Python :: 2',
         'Programming Language :: Python :: 2.7',
+        'Programming Language :: Python :: 3',
+        'Programming Language :: Python :: 3.5',
```
Comment (RevolutionTech) [inline setup.py]: Would you mind adding 3.4 and 3.6 here as well, if those are working?

File: setup.py [hunk 4]
```diff
@@ -43,5 +36,12 @@
...
         'weighted_levenshtein': ['clev.pxd', 'clev.pyx']
     },
-    ext_modules=extensions,
+    
+    setup_requires=[
+        # Setuptools 18.0 properly handles Cython extensions.
+        'setuptools>=18.0',
+        'cython',
+    ],   
+
```
Comment (RevolutionTech) [inline setup.py]: nit: Remove blank newline

File: setup.py [hunk 5]
```diff
@@ -43,5 +36,12 @@
     package_data={
         'weighted_levenshtein': ['clev.pxd', 'clev.pyx']
     },
-    ext_modules=extensions,
+
```
Comment (RevolutionTech) [inline setup.py]: nit: Remove blank newline

File: tox.ini [hunk 6]
```diff
@@ -1,5 +1,6 @@
 [tox]
-envlist = py27
+envlist = py27, py34, py35, py36
+skip_missing_interpreters = true
```
Comment (RevolutionTech) [inline tox.ini]: I don't think we should use this option. Wouldn't that cause CircleCI to pass tests even if it never tested some of the Python versions?

File: weighted_levenshtein/__init__.py [hunk 7]
```diff
@@ -1 +1,2 @@
-from clev import *
+from .clev import *
+
```
Comment (RevolutionTech) [inline weighted_levenshtein/__init__.py]: nit: Remove blank newline at the EOF (the last character of the file should still be a newline, just not the last two characters)

File: weighted_levenshtein/clev.pyx [hunk 8]
```diff
@@ -1,22 +1,12 @@
...
-cdef DTYPE_t unit_array[ALPHABET_SIZE]
-cdef DTYPE_t unit_matrix[ALPHABET_SIZE][ALPHABET_SIZE]
-
-for ii in range(ALPHABET_SIZE):
-    unit_array[ii] = 1
-
-for ii in range(ALPHABET_SIZE):
-    for jj in range(ALPHABET_SIZE):
-        unit_matrix[ii][jj] = 1
+import numpy as np
```
Comment (RevolutionTech) [inline weighted_levenshtein/clev.pyx]: Prior to this change, `numpy` was only a requirement for the package to run tests, but not actually to use any of the functions in the `weighted-levenshtein` package. Can we still use the custom-defined unit array and unit matrix for Python 3 to prevent adding `numpy` as a requirement for the package?

File: weighted_levenshtein/clev.pyx [hunk 9]
```diff
@@ -32,7 +22,7 @@ ctypedef struct Array2D:
 
 cdef inline void Array2D_init(
     Array2D* array2d,
-    Py_ssize_t num_rows,
+    Py_ssize_t num_rows,
```
Comment (RevolutionTech) [inline weighted_levenshtein/clev.pyx]: nit: Please do not add in these unnecessary spaces (also applies to many of the lines below in this file)

File: weighted_levenshtein/clev.pyx [hunk 10]
```diff
@@ -92,43 +82,43 @@ cdef inline DTYPE_t* Array2D_0_at(
...
     Calculates the cost incurred by inserting
     characters 'start' to 'end' (inclusive) from 'str2',
     assuming that 'str2' is 1-indexed.
 
-    Works since row 0 of 'd' is the cumulative sums
+    Works since row 0 of 'd' is the cumulative sums 
     of the insertion costs of the characters in str2.
-
-    This function computes the range sum by computing the difference
+
```
Comment (RevolutionTech) [inline weighted_levenshtein/clev.pyx]: nit: Do not add unnecessary spaces

File: weighted_levenshtein/clev.pyx [hunk 11]
```diff
@@ -456,4 +455,4 @@ cdef DTYPE_t c_levenshtein(
 
     ret_val = Array2D_0_get(d, len1, len2)
     Array2D_del(d)
-    return ret_val
+    return ret_val
```
Comment (RevolutionTech) [inline weighted_levenshtein/clev.pyx]: nit: Please end the file with a newline

File: weighted_levenshtein/clev.pyx [hunk 12]
```diff
@@ -440,7 +439,7 @@ cdef DTYPE_t c_levenshtein(
     for j in range(1, len2 + 1):
         char_j = str_1_get(str2, j)
         Array2D_0_at(d, 0, j)[0] = Array2D_0_get(d, 0, j - 1) + insert_costs[char_j]
-
+
```
Comment (RevolutionTech) [inline weighted_levenshtein/clev.pyx]: nit: Please remove unnecessary whitespace

**Review by RevolutionTech — CHANGES_REQUESTED** [review]
Thanks for your contribution! The PR mostly looks good.

I am wondering if we can get away without adding `numpy` as a required dependency and if we should remove `skip_missing_interpreters` from the `tox.ini`.

Finally, since you've put the work into adding Python 3 support, let's update the docs to say so! 😃 You'll want to update the line of the README that says "Python 3 compatibility is untested".

File: setup.py (hunk 3 above)
Comment (duxan) [reply setup.py]: I have tested only on 3.6, so I'll put 3.6 here and in tox.ini too, as sole py3 versions - that way we keep latest versions only. If you think previous ones should also be included/tested, I'll add them too.

File: weighted_levenshtein/clev.pyx (hunk 8 above)
Comment (duxan) [reply weighted_levenshtein/clev.pyx]: Done. I can't remember why I needed this during the initial changes. Tested now without and it works OK.

**Comment by duxan** [comment]
Hi @RevolutionTech, 
I've updated the changes requested. Would you want me to squash the commits into single one before merge?

**Comment by rampeer** [comment]
I've tried your changes with Python 3.5 (Python 3.5.2+ (default, Sep 22 2016, 12:18:14)):

```
import numpy as np
from weighted_levenshtein import lev, osa, dam_lev

iw = np.ones(128, dtype=np.float64)
dw = np.ones(128, dtype=np.float64)
sw = np.ones((128, 128), dtype=np.float64)
tw = np.ones((128, 128), dtype=np.float64)
print(lev('1234', '124', iw, dw, sw))

insert_costs = np.ones(128, dtype=np.float64)
print(lev('1234', '124', insert_costs=insert_costs))
```

First print outputs 1.0, as expected. Second one fails with 

```
     10 
     11 insert_costs = np.ones(128, dtype=np.float64)
---> 12 print(lev('1234', '124', insert_costs=insert_costs))

/usr/local/lib/python3.5/dist-packages/weighted_levenshtein/clev.pyx in weighted_levenshtein.clev.levenshtein (weighted_levenshtein/clev.c:4553)()

TypeError: expected bytes, str found
```

It seems that something is wrong with default values, or am I missing something?

Also, would you kindly consider updating examples in README? Python2-style `print`s are not copy-paste friendly.

**Comment by duxan** [comment]
@rampeer You are right: default values are not properly set when running with python3, and now I remember why I used numpy in clev.pyx :). I still don't have other fix than using numpy, but for now I added tests for this case when there is no initialization of costs.

File: test/test.py [hunk 13]
```diff
@@ -152,6 +152,53 @@ def test_dl_transpose4(self):
         self.assertEqual(self._dl('bca', 'ab'), 2.5)
 
 
+class TestClev_noinit(unittest.TestCase):
+
+    def setUp(self):
```
Comment (RevolutionTech) [inline test/test.py]: Since you're not doing anything in `setUp()`, this method isn't necessary.

File: test/test.py [hunk 14]
```diff
@@ -152,6 +152,53 @@ def test_dl_transpose4(self):
...
+        pass
+
+    def _lev(self, x, y):
```
Comment (RevolutionTech) [inline test/test.py]: Why use this method? Why not just call `lev` directly?

File: test/test.py [hunk 15]
```diff
@@ -152,6 +152,53 @@ def test_dl_transpose4(self):
...
+        return lev(x, y)
+
+    def _osa(self, x, y):
```
Comment (RevolutionTech) [inline test/test.py]: Same as above. Why not just use `osa` directly? Also with `dam_lev` below.

File: test/test.py [hunk 16]
```diff
@@ -152,6 +152,53 @@ def test_dl_transpose4(self):
         self.assertEqual(self._dl('bca', 'ab'), 2.5)
 
 
+class TestClev_noinit(unittest.TestCase):
```
Comment (RevolutionTech) [inline test/test.py]: Classes in Python usually shouldn't have underscores in the name. I'm also not sure that `noinit` communicates what's going on in this test as well as something like `TestClevUsingDefaultValues`.

File: test/test.py [hunk 17]
```diff
@@ -152,6 +152,53 @@ def test_dl_transpose4(self):
...
+    def _lev(self, x, y):
+        return lev(x, y)
+
+    def _osa(self, x, y):
+        return osa(x, y)
+
+    def _dl(self, x, y):
+        return dam_lev(x, y)
+
+    def test_lev_noinit(self):
```
Comment (RevolutionTech) [inline test/test.py]: If we rename the class above, don't forget to rename this test to correspond with it. It would also be acceptable to just rename this test method to `test_lev`. Same goes with the remaining test methods.

File: test/test.py [hunk 18]
```diff
@@ -152,6 +152,53 @@ def test_dl_transpose4(self):
...
+        self.assertEqual(self._dl('a', 'b'), 1)
+        self.assertEqual(self._dl('a', 'ab'), 1)
+        self.assertEqual(self._dl('ab', 'ba'), 1)
+        self.assertEqual(self._dl('ab', 'bca'), 2)
+        self.assertEqual(self._dl('bca', 'ab'), 2)
+        self.assertEqual(self._dl('ab', 'bdca'), 3)
+        self.assertEqual(self._dl('bdca', 'ab'), 3)
+
+
 if __name__ == '__main__':
```
Comment (RevolutionTech) [inline test/test.py]: Since `python setup.py test` and `tox` work now, I wonder if it makes sense to just remove this logic entirely.

File: weighted_levenshtein/.gitignore [hunk 19]
```diff
@@ -0,0 +1 @@
+clev.c
```
Comment (RevolutionTech) [inline weighted_levenshtein/.gitignore]: nit: Add newline to the EOF

File: weighted_levenshtein/__init__.py [hunk 20]
```diff
@@ -1 +1 @@
-from clev import *
+from .clev import *
```
Comment (RevolutionTech) [inline weighted_levenshtein/__init__.py]: nit: Add newline to the EOF

I realize that previously I asked you to remove the newline at the EOF, but it looks like you've removed two newlines. Each file should end with exactly one newline, no more, no less. So for example:
```
import os
```
is wrong
```
import os\n\n
```
is also wrong, but
```
import os\n
```
is correct. 👌 

I apologize for being so nit-picky, it's probably super annoying.

**Review by RevolutionTech — CHANGES_REQUESTED** [review]
Thanks for all of your help so far, @duxan! I think we'll definitely squash and merge your changes once everything is good to go but I am happy to do that or you can do that.

The PR is looking good, I did have some feedback regarding the new TestCase you created, but it should be pretty straightforward.

Do you know why the default values are not being properly set in Python 3? It's not obvious to me why that wouldn't work by looking at it. If you're not sure I could dig in a little bit to find out, but I might not be able to get to it until next week.

@rampeer makes a good suggestion in that the README should probably be updated so that the print statements are Python 3-friendly. Would you be able to make that edit as well?

File: weighted_levenshtein/__init__.py (hunk 20 above)
Comment (duxan) [reply weighted_levenshtein/__init__.py]: no worries :)

File: weighted_levenshtein/.gitignore (hunk 19 above)
Comment (duxan) [reply weighted_levenshtein/.gitignore]: added, not shown on github, but says 2 lines

**Comment by duxan** [comment]
I guess python 3 errors have something to do with how unit_array and unit_matrix were constructed, since when using cython arrays everything works OK (similar as with numpy arrays). This problem is definitely beyond my reach. I have used this doc as a reference of how to construct cython arrays: http://docs.cython.org/en/latest/src/userguide/memoryviews.html#quickstart and there is also a benchmark between methods. It seems worth investigating this further and finding most optimal way for this use-case. 

Another thing here is that some use-cases would benefit from having ALPHABET_SIZE configurable, like in bioinformatics (for sequence alignment/comparison) where alphabet is much smaller.

**Comment by pachewise** [comment]
It's an issue with how Cython parses types and the `c_string_type=str` setting meaning something different for Python 3.

clev.c:4719
````c
    /* "weighted_levenshtein/clev.pyx":410
 *         insert_costs = unit_array
 *     if delete_costs is None:
 *         delete_costs = unit_array             # <<<<<<<<<<<<<<
 *     if substitute_costs is None:
 *         substitute_costs = unit_matrix
 */
    __Pyx_TraceLine(410,0,__PYX_ERR(0, 410, __pyx_L1_error))
    __pyx_t_3 = __pyx_format_from_typeinfo(&__Pyx_TypeInfo_nn___pyx_t_20weighted_levenshtein_4clev_DTYPE_t);
    __pyx_t_4 = Py_BuildValue((char*) "("  __PYX_BUILD_PY_SSIZE_T  ")", ((Py_ssize_t)__pyx_e_20weighted_levenshtein_4clev_ALPHABET_SIZE));
    if (unlikely(!__pyx_t_3 || !__pyx_t_4 || !PyBytes_AsString(__pyx_t_3))) __PYX_ERR(0, 410, __pyx_L1_error)
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_2 = __pyx_array_new(__pyx_t_4, sizeof(__pyx_t_20weighted_levenshtein_4clev_DTYPE_t), PyBytes_AS_STRING(__pyx_t_3), (char *) "fortran", (char *) __pyx_v_20weighted_levenshtein_4clev_unit_array);
````

I have a hunch that`format_from_typeinfo` is returning `str` (unicode) data, but haven't been able to confirm. Changing to `c_string_type=bytes` fixes 
@rampeer's usecase when using the previous `unit_*` impl, but breaks 3.6 tests (rounding errors).

Regardless, Numpy arrays or cython arrays seem to be the right choice for `unit_*` (look at the last line above - they were being converted to cython arrays anyway).

**Review by RevolutionTech — APPROVED** [review]
@duxan Thank you so much for all of your hard work! All of us at InfoScout greatly appreciate your contribution and I know it will make many happy to have Python 3 compatibility in this library now. 😃 

Making `ALPHABET_SIZE` configurable sounds like a good idea to me and if it would be useful to you I would welcome a future PR to add that functionality.

Thanks again!

**Comment by mathiasose** [comment]
Don't forget to upload the new version to PyPI.

**Comment by RevolutionTech** [comment]
@dsu1995 I have created a new release tag v0.2. Would you be able to upload this version to PyPi?
